File: core/src/main/scala/eu/ostrzyciel/jelly/core/ProtoTranscoderImpl.scala
package eu.ostrzyciel.jelly.core

import eu.ostrzyciel.jelly.core.proto.v1.*

import scala.annotation.switch
import scala.collection.mutable.ArrayBuffer

// Note: the caller is responsible for setting a physical stream type in the output!

private final class ProtoTranscoderImpl(
  supportedInputOptions: Option[RdfStreamOptions],
  outputOptions: RdfStreamOptions
) extends ProtoTranscoder:
  override def ingestRow(row: RdfStreamRow): Iterable[RdfStreamRow] =
    rowBuffer.clear()
    processRow(row)
    rowBuffer.toSeq

  override def ingestFrame(frame: RdfStreamFrame): RdfStreamFrame =
    rowBuffer.clear()
    for row <- frame.rows do
      processRow(row)
    val newFrame = RdfStreamFrame(rowBuffer.toSeq)
    newFrame

  // Transcoder state
  private val checkInputOptions = supportedInputOptions.isDefined
  private val prefixLookup: TranscoderLookup = new TranscoderLookup(false, outputOptions.maxPrefixTableSize)
  private val nameLookup: TranscoderLookup = new TranscoderLookup(true, outputOptions.maxNameTableSize)
  private val datatypeLookup: TranscoderLookup = new TranscoderLookup(false, outputOptions.maxDatatypeTableSize)

  // Current input stream state
  private var inputOptions: RdfStreamOptions = null
  private var inputUsesPrefixes = false

  // Current output stream state
  private val rowBuffer = new ArrayBuffer[RdfStreamRow](128)
  private var changeInTerms = false
  private var emittedOptions = false

  private def processRow(row: RdfStreamRow): Unit =
    val r = row.row
    if r == null then
      throw new RdfProtoTranscodingError("Row kind is not set.")
    (r.streamRowValueNumber: @switch) match
      case RdfStreamRow.OPTIONS_FIELD_NUMBER => handleOptions(r.options)
      case RdfStreamRow.TRIPLE_FIELD_NUMBER => handleTriple(row)
      case RdfStreamRow.QUAD_FIELD_NUMBER => handleQuad(row)
      case RdfStreamRow.GRAPH_START_FIELD_NUMBER =>
        this.changeInTerms = false
        val term = r.graphStart.graph
        val g1 = handleGraphTerm(term)
        if changeInTerms then
          rowBuffer.append(RdfStreamRow(RdfGraphStart(g1)))
        else rowBuffer.append(row)
      case RdfStreamRow.GRAPH_END_FIELD_NUMBER => rowBuffer.append(row)
      case RdfStreamRow.NAME_FIELD_NUMBER =>
        val name = r.name
        val entry = nameLookup.addEntry(name.id, name.value)
        if !entry.newEntry then ()
        else if entry.setId == name.id then rowBuffer.append(row)
        else rowBuffer.append(RdfStreamRow(RdfNameEntry(entry.setId, name.value)))
      case RdfStreamRow.PREFIX_FIELD_NUMBER =>
        val prefix = r.prefix
        val entry = prefixLookup.addEntry(prefix.id, prefix.value)
        if !entry.newEntry then ()
        else if entry.setId == prefix.id then rowBuffer.append(row)
        else rowBuffer.append(RdfStreamRow(RdfPrefixEntry(entry.setId, prefix.value)))
      case RdfStreamRow.DATATYPE_FIELD_NUMBER =>
        val datatype = r.datatype
        val entry = datatypeLookup.addEntry(datatype.id, datatype.value)
        if !entry.newEntry then ()
        else if entry.setId == datatype.id then rowBuffer.append(row)
        else rowBuffer.append(RdfStreamRow(RdfDatatypeEntry(entry.setId, datatype.value)))
      case _ =>
        // This case should never happen
        throw new RdfProtoTranscodingError("Row kind is not set.")

  private def handleTriple(row: RdfStreamRow): Unit =
    this.changeInTerms = false
    val triple = row.row.triple
    val s1 = handleSpoTerm(triple.subject)
    val p1 = handleSpoTerm(triple.predicate)
    val o1 = handleSpoTerm(triple.`object`)
    if changeInTerms then
      rowBuffer.append(RdfStreamRow(RdfTriple(s1, p1, o1)))
    else rowBuffer.append(row)

  private def handleQuad(row: RdfStreamRow): Unit =
    this.changeInTerms = false
    val quad = row.row.quad
    val s1 = handleSpoTerm(quad.subject)
    val p1 = handleSpoTerm(quad.predicate)
    val o1 = handleSpoTerm(quad.`object`)
    val g1 = handleGraphTerm(quad.graph)
    if changeInTerms then
      rowBuffer.append(RdfStreamRow(RdfQuad(s1, p1, o1, g1)))
    else rowBuffer.append(row)

  private def handleSpoTerm(term: SpoTerm): SpoTerm =
    if term == null then null
    else if term.isIri then handleIri(term.iri)
    else if term.isBnode then term
    else if term.isLiteral then handleLiteral(term.literal)
    else if term.isTripleTerm then handleTripleTerm(term.tripleTerm)
    else throw new RdfProtoTranscodingError("Unknown term type.")

  private def handleGraphTerm(term: GraphTerm): GraphTerm =
    if term == null then null
    else if term.isIri then handleIri(term.iri)
    else if term.isDefaultGraph then term
    else if term.isBnode then term
    else if term.isLiteral then handleLiteral(term.literal)
    else throw new RdfProtoTranscodingError("Unknown term type.")

  private def handleIri(iri: RdfIri): RdfIri =
    val prefix = iri.prefixId
    val name = iri.nameId
    val prefix1 = if inputUsesPrefixes then prefixLookup.remap(prefix) else 0
    val name1 = nameLookup.remap(name)
    if prefix1 != prefix || name1 != name then
      changeInTerms = true
      RdfIri(prefix1, name1)
    else iri

  private def handleLiteral(literal: RdfLiteral): RdfLiteral =
    if !literal.literalKind.isDatatype then literal
    else
      val dt = literal.literalKind.datatype
      val dt1 = datatypeLookup.remap(dt)
      if dt1 != dt then
        changeInTerms = true
        RdfLiteral(literal.lex, RdfLiteral.LiteralKind.Datatype(dt1))
      else literal

  private def handleTripleTerm(triple: RdfTriple): RdfTriple =
    val s1 = handleSpoTerm(triple.subject)
    val p1 = handleSpoTerm(triple.predicate)
    val o1 = handleSpoTerm(triple.`object`)
    if s1 != triple.subject || p1 != triple.predicate || o1 != triple.`object` then
      changeInTerms = true
      RdfTriple(s1, p1, o1)
    else triple

  private def handleOptions(options: RdfStreamOptions): Unit =
    if checkInputOptions then
      if outputOptions.physicalType != options.physicalType then
        throw new RdfProtoTranscodingError("Input stream has a different physical type than the output. " +
          f"Input: ${options.physicalType} output: ${outputOptions.physicalType}")
      JellyOptions.checkCompatibility(options, supportedInputOptions.get)
    this.inputUsesPrefixes = options.maxPrefixTableSize > 0
    if inputUsesPrefixes then
      prefixLookup.newInputStream(options.maxPrefixTableSize)
    else if outputOptions.maxPrefixTableSize > 0 then
      throw new RdfProtoTranscodingError("Output stream uses prefixes, but the input stream does not.")
    nameLookup.newInputStream(options.maxNameTableSize)
    datatypeLookup.newInputStream(options.maxDatatypeTableSize)
    // Update the input options
    inputOptions = options
    if !emittedOptions then
      emittedOptions = true
      rowBuffer.append(RdfStreamRow(outputOptions.copy(
        version = Constants.protoVersion
      )))

File: core/src/main/java/eu/neverblink/jelly/core/internal/ProtoTranscoderImpl.java
package eu.neverblink.jelly.core.internal;

import eu.neverblink.jelly.core.*;
import eu.neverblink.jelly.core.proto.v1.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Fast implementation of the ProtoTranscoder interface.
 * <p>
 * It does not in perfect compression (like you would get with full decoding and re-encoding), but it should be
 * good enough for the vast majority of cases.
 */
@InternalApi
public final class ProtoTranscoderImpl implements ProtoTranscoder {

    private final RdfStreamOptions supportedInputOptions;
    private final RdfStreamOptions outputOptions;

    private final TranscoderLookup prefixLookup;
    private final TranscoderLookup nameLookup;
    private final TranscoderLookup datatypeLookup;

    private final List<RdfStreamRow> rowBuffer = new ArrayList<>();

    private RdfStreamOptions inputOptions = null;
    private boolean inputUsesPrefixes = false;
    private boolean hasChangedTerms = false;
    private boolean hasEmittedOptions = false;

    /**
     * Constructor for the ProtoTranscoderImpl class.
     *
     * @param supportedInputOptions maximum allowable options for the input streams (optional)
     * @param outputOptions options for the output stream. This MUST have the physical stream type set.
     */
    public ProtoTranscoderImpl(RdfStreamOptions supportedInputOptions, RdfStreamOptions outputOptions) {
        this.supportedInputOptions = supportedInputOptions;
        this.outputOptions = outputOptions;
        prefixLookup = new TranscoderLookup(false, outputOptions.getMaxPrefixTableSize());
        nameLookup = new TranscoderLookup(true, outputOptions.getMaxNameTableSize());
        datatypeLookup = new TranscoderLookup(false, outputOptions.getMaxDatatypeTableSize());
    }

    @Override
    public Iterable<RdfStreamRow> ingestRow(RdfStreamRow row) {
        rowBuffer.clear();
        processRow(row);
        return rowBuffer;
    }

    @Override
    public RdfStreamFrame ingestFrame(RdfStreamFrame frame) {
        rowBuffer.clear();
        for (final var row : frame.getRows()) {
            processRow(row);
        }
        final var outFrame = RdfStreamFrame.newInstance();
        outFrame.getRows().addAll(rowBuffer);
        outFrame.getMetadata().addAll(frame.getMetadata());
        return outFrame;
    }

    private void processRow(RdfStreamRow row) {
        switch (row.getRowFieldNumber()) {
            case RdfStreamRow.OPTIONS -> handleOptions(row.getOptions());
            case RdfStreamRow.TRIPLE -> handleTriple(row);
            case RdfStreamRow.QUAD -> handleQuad(row);
            case RdfStreamRow.GRAPH_START -> handleGraphStart(row);
            case RdfStreamRow.GRAPH_END -> handleIdentity(row);
            case RdfStreamRow.NAMESPACE -> handleNamespaceDeclaration(row);
            case RdfStreamRow.NAME -> handleName(row);
            case RdfStreamRow.PREFIX -> handlePrefix(row);
            case RdfStreamRow.DATATYPE -> handleDatatype(row);
            default -> throw new RdfProtoTranscodingError("Row kind is not set");
        }
    }

    private void handleName(RdfStreamRow row) {
        final var name = row.getName();
        final var entry = nameLookup.addEntry(name.getId(), name.getValue());
        if (!entry.newEntry) {
            return;
        }

        if (entry.setId == name.getId()) {
            rowBuffer.add(row);
            return;
        }

        final var newName = RdfNameEntry.newInstance().setId(entry.setId).setValue(name.getValue());
        rowBuffer.add(RdfStreamRow.newInstance().setName(newName));
    }

    private void handlePrefix(RdfStreamRow row) {
        final var prefix = row.getPrefix();
        final var entry = prefixLookup.addEntry(prefix.getId(), prefix.getValue());
        if (!entry.newEntry) {
            return;
        }

        if (entry.setId == prefix.getId()) {
            rowBuffer.add(row);
            return;
        }

        final var newPrefix = RdfPrefixEntry.newInstance().setId(entry.setId).setValue(prefix.getValue());
        rowBuffer.add(RdfStreamRow.newInstance().setPrefix(newPrefix));
    }

    private void handleDatatype(RdfStreamRow row) {
        final var datatype = row.getDatatype();
        final var entry = datatypeLookup.addEntry(datatype.getId(), datatype.getValue());
        if (!entry.newEntry) {
            return;
        }

        if (entry.setId == datatype.getId()) {
            rowBuffer.add(row);
            return;
        }

        final var newDatatype = RdfDatatypeEntry.newInstance().setId(entry.setId).setValue(datatype.getValue());
        rowBuffer.add(RdfStreamRow.newInstance().setDatatype(newDatatype));
    }

    private void handleIdentity(RdfStreamRow row) {
        // No changes needed, just add the row to the buffer
        rowBuffer.add(row);
    }

    private void handleTriple(RdfStreamRow row) {
        this.hasChangedTerms = false;
        final var triple = row.getTriple();

        final var s1 = handleSpoTerm(triple.getSubject());
        final var p1 = handleSpoTerm(triple.getPredicate());
        final var o1 = handleSpoTerm(triple.getObject());

        if (!hasChangedTerms) {
            rowBuffer.add(row);
            return;
        }

        final var newTriple = RdfTriple.newInstance().setSubject(s1).setPredicate(p1).setObject(o1);
        rowBuffer.add(RdfStreamRow.newInstance().setTriple(newTriple));
    }

    private void handleQuad(RdfStreamRow row) {
        this.hasChangedTerms = false;
        final var quad = row.getQuad();

        final var s1 = handleSpoTerm(quad.getSubject());
        final var p1 = handleSpoTerm(quad.getPredicate());
        final var o1 = handleSpoTerm(quad.getObject());
        final var g1 = handleGraphTerm(quad.getGraph());

        if (!hasChangedTerms) {
            rowBuffer.add(row);
            return;
        }

        final var newQuad = RdfQuad.newInstance().setSubject(s1).setPredicate(p1).setObject(o1).setGraph(g1);
        rowBuffer.add(RdfStreamRow.newInstance().setQuad(newQuad));
    }

    private void handleGraphStart(RdfStreamRow row) {
        this.hasChangedTerms = false;
        final var graphStart = row.getGraphStart();

        final var g1 = handleGraphTerm(graphStart.getGraph());
        if (!hasChangedTerms) {
            rowBuffer.add(row);
            return;
        }

        final var newGraphStart = RdfGraphStart.newInstance();
        newGraphStart.setGraph(g1);
        rowBuffer.add(RdfStreamRow.newInstance().setGraphStart(newGraphStart));
    }

    private void handleNamespaceDeclaration(RdfStreamRow row) {
        this.hasChangedTerms = false;
        var nsRow = row.getNamespace();
        var iriValue = handleIri(nsRow.getValue());

        if (!hasChangedTerms) {
            rowBuffer.add(row);
            return;
        }

        var namespace = RdfNamespaceDeclaration.newInstance().setName(nsRow.getName()).setValue(iriValue);

        rowBuffer.add(RdfStreamRow.newInstance().setNamespace(namespace));
    }

    private Object handleSpoTerm(Object term) {
        if (term == null) {
            return null;
        } else if (term instanceof RdfIri.Mutable iri) {
            return handleIri(iri);
        } else if (term instanceof String blankNode) {
            return blankNode;
        } else if (term instanceof RdfLiteral.Mutable literal) {
            return handleLiteral(literal);
        } else if (term instanceof RdfTriple.Mutable triple) {
            return handleTripleTerm(triple);
        } else {
            throw new RdfProtoTranscodingError("Unknown term type");
        }
    }

    private Object handleGraphTerm(Object graph) {
        if (graph == null) {
            return null;
        } else if (graph instanceof RdfIri.Mutable iri) {
            return handleIri(iri);
        } else if (graph instanceof String blankNode) {
            return blankNode; // blank node
        } else if (graph instanceof RdfDefaultGraph.Mutable) {
            return graph; // default graph
        } else if (graph instanceof RdfLiteral.Mutable literal) {
            return handleLiteral(literal);
        } else {
            throw new RdfProtoTranscodingError("Unknown graph term type");
        }
    }

    private RdfIri handleIri(RdfIri iri) {
        var prefix = iri.getPrefixId();
        var name = iri.getNameId();
        var prefix1 = inputUsesPrefixes ? prefixLookup.remap(prefix) : 0;
        var name1 = nameLookup.remap(name);
        if (prefix1 != prefix || name1 != name) {
            hasChangedTerms = true;
            return RdfIri.newInstance().setPrefixId(prefix1).setNameId(name1);
        }
        return iri;
    }

    private RdfLiteral handleLiteral(RdfLiteral literal) {
        if (!literal.hasDatatype()) {
            return literal;
        }

        var dt = literal.getDatatype();
        var dt1 = datatypeLookup.remap(dt);
        if (dt1 != dt) {
            hasChangedTerms = true;
            return RdfLiteral.newInstance().setLex(literal.getLex()).setDatatype(dt1);
        }

        return literal;
    }

    private RdfTriple handleTripleTerm(RdfTriple triple) {
        var s1 = handleSpoTerm(triple.getSubject());
        var p1 = handleSpoTerm(triple.getPredicate());
        var o1 = handleSpoTerm(triple.getObject());
        // Reference comparison is fine here, as the term objects should be reused directly if possible.
        if (s1 != triple.getSubject() || p1 != triple.getPredicate() || o1 != triple.getObject()) {
            hasChangedTerms = true;
            return RdfTriple.newInstance().setSubject(s1).setPredicate(p1).setObject(o1);
        }
        return triple;
    }

    private void handleOptions(RdfStreamOptions options) {
        if (supportedInputOptions != null) {
            if (outputOptions.getPhysicalType() != options.getPhysicalType()) {
                throw new RdfProtoTranscodingError(
                    "Input stream has a different physical type than the output. Input: %s output: %s".formatted(
                        options.getPhysicalType(),
                        outputOptions.getPhysicalType()
                    )
                );
            }
            JellyOptions.checkCompatibility(options, supportedInputOptions);
        }

        this.inputUsesPrefixes = options.getMaxPrefixTableSize() > 0;

        if (inputUsesPrefixes) {
            prefixLookup.newInputStream(options.getMaxPrefixTableSize());
        } else if (outputOptions.getMaxPrefixTableSize() > 0) {
            throw new RdfProtoTranscodingError("Output stream uses prefixes, but the input stream does not.");
        }

        nameLookup.newInputStream(options.getMaxNameTableSize());
        datatypeLookup.newInputStream(options.getMaxDatatypeTableSize());

        // Set the input options
        inputOptions = options;

        // Update the input options
        if (hasEmittedOptions) {
            return;
        }

        hasEmittedOptions = true;
        var version =
            inputOptions.getVersion() == JellyConstants.PROTO_VERSION_1_0_X
                ? JellyConstants.PROTO_VERSION_1_0_X
                : JellyConstants.PROTO_VERSION;

        var newOptions = outputOptions.clone().setVersion(version);
        rowBuffer.add(RdfStreamRow.newInstance().setOptions(newOptions));
    }
}
